hypershift-operator/controllers/openshiftcluster_controller.go

```diff
@@ -18,164 +18,199 @@ package controllers
-	if err != nil {
-		r.Log.Error(err, "failed to ensure infrastructure")
-		return result, fmt.Errorf("failed to ensure infrastructure: %w", err)
+	hcp := &hyperv1.HostedControlPlane{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: ocluster.GetNamespace(),
+			Name:      ocluster.GetName(),
+		},
+		Spec: hyperv1.HostedControlPlaneSpec{
+			BaseDomain:   ocluster.Spec.BaseDomain,
+			PullSecret:   ocluster.Spec.PullSecret,
+			ServiceCIDR:  ocluster.Spec.ServiceCIDR,
+			PodCIDR:      ocluster.Spec.PodCIDR,
+			SSHKey:       ocluster.Spec.SSHKey,
+			ReleaseImage: ocluster.Spec.ReleaseImage,
+		},
+	}
+	guestCluster := &hyperv1.GuestCluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: ocluster.GetNamespace(),
+			Name:      ocluster.GetName(),
+		},
+		Spec: hyperv1.GuestClusterSpec{
+			ComputeReplicas: ocluster.Spec.ComputeReplicas,
+		},
 	}
 
-	// Wait for things like LB services to become available
-	if !infraStatus.IsReady() {
-		result.RequeueAfter = 5 * time.Second
-		r.Log.Info("cluster infrastructure is still provisioning, will try again later")
-		return result, nil
+	if _, err := ctrl.CreateOrUpdate(ctx, r.Client, cluster, func() error { return nil }); err != nil {
+		return ctrl.Result{}, err
 	}
 
-	// Install the control plane into the infrastructure
-	err = r.ensureControlPlane(ctx, cluster, infraStatus, &releaseInfo)
-	if err != nil {
-		r.Log.Error(err, "failed to ensure control plane")
-		return result, fmt.Errorf("failed to ensure control plane: %w", err)
+	if _, err := ctrl.CreateOrUpdate(ctx, r.Client, hcp, func() error { return nil }); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if _, err := ctrl.CreateOrUpdate(ctx, r.Client, guestCluster, func() error { return nil }); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	var currentCluster capiv1.Cluster
+	if err := r.Get(ctx, client.ObjectKeyFromObject(cluster), &currentCluster); err != nil {
+		return ctrl.Result{}, err
 	}
+	r.Log = r.Log.WithValues("cluster", cluster.Name)
 
+	ready := currentCluster.Status.ControlPlaneReady && currentCluster.Status.InfrastructureReady
+	if !ready {
+		r.Log.Info("Not ready yet. Requeueing")
+		return ctrl.Result{Requeue: true}, nil
+	}
+	ocluster.Status.Ready = ready
+	if err := patchHelper.Patch(ctx, ocluster); err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to patch: %w", err)
+	}
+
+	r.Log.Info("Successfully reconciled")
 	return ctrl.Result{}, nil
 }
 
-func (r *OpenShiftClusterReconciler) delete(ctx context.Context, name string) error {
-	ns := &corev1.Namespace{
-		ObjectMeta: metav1.ObjectMeta{Name: name},
+func (r *OpenShiftClusterReconciler) delete(ctx context.Context, name, namespace string) error {
+	cluster := &capiv1.Cluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+		},
 	}
-	if err := r.Delete(ctx, ns); err != nil && !apierrors.IsNotFound(err) {
-		return fmt.Errorf("failed to delete namespace: %w", err)
+	hcp := &hyperv1.HostedControlPlane{
```

Comment:
If connected via owner refs, is it necessary to delete the hcp and guestcluster?